ciri/consensus/snapshot/local_snapshots/pruning_service.c

```diff
@@ -0,0 +1,333 @@
+  }
+
+cleanup:
+  if (err != RC_OK) {
+    params.current_snapshot_index = ps->last_pruned_snapshot_index;
+    hash243_set_free(params.transactions_to_prune);
+  }
+  return RC_OK;
+}
+
+static retcode_t prune_transactions(pruning_service_t *const ps, tangle_t const *const tangle,
+                                    spent_addresses_provider_t *const sap) {
+  retcode_t err;
+  DECLARE_PACK_SINGLE_MILESTONE(milestone, milestone_ptr, milestone_pack);
+  hash243_set_t transactions_to_prune = NULL;
+
+  ERR_BIND_GOTO(iota_tangle_milestone_load_by_index(tangle, ps->last_pruned_snapshot_index, &milestone_pack), err,
+                cleanup);
+
+  ERR_BIND_GOTO(collect_transactions_to_prune(ps, tangle, sap, milestone.hash, &transactions_to_prune), err, cleanup);
+  ERR_BIND_GOTO(iota_tangle_transaction_delete(tangle, transactions_to_prune), err, cleanup);
+  hash243_set_free(&transactions_to_prune);
+  // It's important to delete the milestone only after all it's past cone has been deleted to avoid dangle transactions
+  hash243_set_add(&transactions_to_prune, milestone.hash);
+  ERR_BIND_GOTO(iota_tangle_milestone_delete(tangle, milestone.hash), err, cleanup);
+  log_info(logger_id,
+           "Snapshot index % " PRIu64 " was pruned successfully, % " PRIu64 " transactions were removed from db\n",
+           ps->last_pruned_snapshot_index, hash243_set_size(transactions_to_prune));
+  ps->last_pruned_snapshot_index++;
+
+  hash243_set_free(&transactions_to_prune);
+
+cleanup:
+
+  if (err != RC_OK) {
+    lock_handle_unlock(&ps->lock_handle);
+    log_warning(logger_id, "Local snapshots pruning has failed with error code: %d\n", err);
+  }
+
+  return RC_OK;
+}
+
+retcode_t iota_local_snapshots_pruning_service_init(pruning_service_t *const ps,
+                                                    snapshots_provider_t *const snapshot_provider,
+                                                    spent_addresses_service_t *const spent_addresses_service,
+                                                    tips_cache_t *const tips_cache,
+                                                    iota_consensus_conf_t const *const conf) {
+  logger_id = logger_helper_enable(PRUNING_SERVICE_LOGGER_ID, LOGGER_DEBUG, true);
+  memset(ps, 0, sizeof(pruning_service_t));
+  ps->conf = conf;
+  ps->last_pruned_snapshot_index = snapshot_provider->inital_snapshot.metadata.index;
+  ps->last_snapshot_index_to_prune = ps->last_pruned_snapshot_index;
+  ps->spent_addresses_service = spent_addresses_service;
+  ps->tips_cache = tips_cache;
+  lock_handle_init(&ps->lock_handle);
+  return RC_OK;
+}
+
+retcode_t iota_local_snapshots_pruning_service_start(pruning_service_t *const ps) {
+  if (ps == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  ps->running = true;
+
+  log_info(logger_id, "Spawning pruning service thread\n");
+  if (thread_handle_create(&ps->pruning_service_thread, (thread_routine_t)pruning_service_routine, ps) != 0) {
```

Comment:
Is it a thread because we expect the pruning to take a long time ? Can't it be done by the LS manager thread itself ?

Reply:
yes, that's the reason, IRI even has jobs for each `milestone`, `unconfirmed transactions branch`,
so i chose the middle

Reply:
Alright